media/virtualDocument.ts

```diff
@@ -134,25 +136,99 @@ export class VirtualDocument {
         // Event handles to handle when the user drags to create a selection
-        editorContainer.addEventListener("mousedown", (click: MouseEvent) => {
-            this.selectHandler.isDragging = true;
-            SelectHandler.selectMouseHandler(click, click.ctrlKey, click.shiftKey);
+        editorContainer.addEventListener("click", (event: MouseEvent) => {
+            const target = event.target as HTMLElement;
+            if (!target || !target.hasAttribute("data-offset")) {
+                return;
+            }
+
+            event.preventDefault();
+
             this.editHandler.completePendingEdits();
-            click.preventDefault();
-        });
-        editorContainer.addEventListener("mousemove", (event: MouseEvent) => {
-            if (event.buttons == 0) this.selectHandler.isDragging = false;
-            if (this.selectHandler.isDragging) {
-                const selected = document.getElementsByClassName("selected") as HTMLCollectionOf<HTMLSpanElement>;
-                let startOffset = selected[selected.length - 1]?.getAttribute("data-offset");
-                const endOffset = (event.target as HTMLSpanElement).getAttribute("data-offset");
-                if (endOffset !== null) {
-                    startOffset = startOffset === null ? endOffset : startOffset;
-                    SelectHandler.multiSelect(createOffsetRange(parseInt(startOffset), parseInt(endOffset)), true);
+
+            if (event.shiftKey) {
+                const focused = this.selectHandler.getFocused();
+                if (focused !== undefined) {
+                    const offset = parseInt(target.getAttribute("data-offset")!);
+
+                    const min = Math.min(focused, offset);
+                    const max = Math.max(focused, offset);
+
+                    this.selectHandler.setSelected(createOffsetRange(min, max));
+
+                    target.focus({ preventScroll: true });
+                }
+            } else {
+                const offset = parseInt(target.getAttribute("data-offset")!);
+
+                this.selectHandler.setFocused(offset);
+
+                if (event.ctrlKey) {
+                    const selection = this.selectHandler.getSelected();
+                    const newSelection = selection.filter(i => i !== offset);
+                    if (selection.length === newSelection.length) {
+                        this.selectHandler.setSelected([...newSelection, offset]);
+                    } else {
+                        this.selectHandler.setSelected(newSelection);
+                    }
+                } else {
+                    this.selectHandler.setSelected([offset]);
                 }
+
+                target.focus({ preventScroll: true });
+
+                this.updateInspector();
             }
         });
-        editorContainer.addEventListener("mouseup", () => this.selectHandler.isDragging = false);
+        editorContainer.addEventListener("mousedown", (event: MouseEvent) => {
```

Comment:
Can we move the logic for these event listeners to functions such as `clickHandler` and `mouseDownHandler` similar to keydown as to not make this already painfully long constructor even longer.